float32 r_data
float32 l_data